### ISSUE TITLE
`Vips::Image#height` of thumbnail_image is different in Ubuntu 22.04

### ISSUE THREAD

**Issue by yahonda** [libvips/ruby-vips#383]
**Describe the bug**
`Vips::Image#height` of thumbnail_image is different in Ubuntu 22.04

**To Reproduce**

1. Install Ruby 3.3.0 on Ubuntu 22.04
```
$ ruby -v
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [aarch64-linux]
```

2. Install the latest `ruby-vips`
```
$ gem install ruby-vips
```

3. Download the webp file 
```
$ wget https://raw.githubusercontent.com/rails/rails/main/activestorage/test/fixtures/files/valley.webp
```

4. Save this file as `thumbnail_height.rb`
```ruby
require 'vips'

# Load the input image
input_image = Vips::Image.new_from_file('valley.webp')

# Define the thumbnail width and height
thumbnail_width = 50
thumbnail_height = 50

# Create the thumbnail image
thumbnail_image = input_image.thumbnail_image(thumbnail_width, height: thumbnail_height)
p thumbnail_image.height
```

5. Run `ruby thumbnail_height.rb`
```
$ ruby thumbnail_height.rb
```

6. It returns `34` as the height of the thumbnail
```
$ ruby thumbnail_height.rb
34
```

**Expected behavior**
It should show `33`

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Ubuntu 22.04.3 LTS
 - Browser `wget` is only used to download the test file.
 - Version `GNU Wget 1.21.2 built on linux-gnu.`

**Additional context**
This issue is opened while investigating Rails CI isssue
https://buildkite.com/rails/rails-nightly/builds/149#018d9052-1b2d-48fa-9d74-a39df3f3f1d6/1251-1291
It had been working fine, that means the height of the thumbnail had been 33.

I attempt to reproduce this issue using libvips42 only but it does not reproduce.

- Steps without ruby-vips
```
$ vipsthumbnail valley.webp -s 50x50
$ identify -format "%h" tn_valley.jpg
33
```

- Installed apt packages
```
$ apt list --installed | grep libvips

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

libvips-doc/jammy,now 8.12.1-1build1 all [installed,automatic]
libvips-tools/jammy,now 8.12.1-1build1 arm64 [installed]
libvips42/jammy,now 8.12.1-1build1 arm64 [installed]
```

**Comment by jcupitt** [comment]
Hello @yahonda,

I would guess you are seeing results from two different libvips versions -- there have been some improvements to the rounding behaviour of `thumbnail`.

This is off-topic, but I would not use `thumbnail_image`, it's only there for emergencies. You will find plain `thumbnail` is often much faster and uses a lot less memory. For example:

```
$ vipsheader nina.jpg
nina.jpg: 6048x4032 uchar, 3 bands, srgb, jpegload
```

I see:

```
$ /usr/bin/time -f %M:%e vips thumbnail_image nina.jpg x.jpg 100
218112:0.25
$ /usr/bin/time -f %M:%e vips thumbnail nina.jpg x.jpg 100
38556:0.06
```

So `thumbnail_image` is 250ms and 220mb of memory, but `thumbnail` is 60ms and 38mb. In fact it's better than that, since startup / shutdown cost for `vips` is included.

```
$ /usr/bin/time -f %M:%e vips > /dev/null
31232:0.02
```

So really it's 230ms and 190mb of memory, versus 40ms and 7mb -- `thumbnail` is 6x faster and 30x less memory.


124 MB of RAM and 0.23s to shink by a factor of 10. With `thumbnail` it's:

```
$ /usr/bin/time -f %M:%e vips thumbnail nina.jpg x.jpg 605
68864:0.08
```

Now it's 68 MB of memory and 0.08s -- half the memory use, and 3x faster. In
fact the improvement is better than that, since the ``vips` command takes a
while to start and needs a fair amount of memory:

```
$ /usr/bin/time -f %M:%e vips > /dev/null
31232:0.02
```

31 MB and 0.02s, so `thumbnail` is really 2.5x less memory and 4x faster.

You can see much larger improvements with other formats, and quality will
often be better as well, since `thumbnail` will automatically premultiply and
can render vector images directly at the correct size.

**Comment by yahonda** [comment]
Thanks for the update.

> I would guess you are seeing results from two different libvips versions -- there have been some improvements to the rounding behaviour of thumbnail.

Thanks for the info. Here are the installed `libvips42` version those height returns `33`.

```
$ apt list --installed | grep libvips

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

libvips-dev/mantic,now 8.14.3-1 amd64 [installed]
libvips-tools/mantic,now 8.14.3-1 amd64 [installed]
libvips42/mantic,now 8.14.3-1 amd64 [installed]
$
```

Here are the installed `libvips42` version those height returns `34`.

```
$ apt list --installed | grep libvips

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

libvips-doc/jammy,now 8.12.1-1build1 all [installed,automatic]
libvips-tools/jammy,now 8.12.1-1build1 arm64 [installed]
libvips42/jammy,now 8.12.1-1build1 arm64 [installed]
```

> This is off-topic, but I would not use thumbnail_image, it's only there for emergencies. You will find plain thumbnail is often much faster and uses a lot less memory. For example:


```
$ vips thumbnail valley.webp thumbnail.jpg 50 --height 50
$ identify -format "%h" thumbnail.jpg
33
```

I'm wondering why even the `vips thumbnail` creates the thumbnail whose height is 33 in command line while ruby-vips returns 34.

**Comment by yahonda** [comment]
I have opened https://github.com/rails/rails/pull/51060 to allow both height, 33 and 34 at Rails CI.

**Comment by jcupitt** [comment]
> I'm wondering why even the vips thumbnail creates the thumbnail whose height is 33 in command line while ruby-vips returns 34.

I think your ruby-vips and vips CLI must be using a different libvips versions.

**Comment by kleisauke** [comment]
This was a bug that has been fixed in libvips 8.13 - see:
https://www.libvips.org/2022/05/28/What's-new-in-8.13.html#image-resize-quality-improvements

Resizing an 550×368 image to a width of 50 pixels should yield a height of 33 pixels.
$`50 \div (550 \div 368) = 33.4545454545 \approx 33`$

**Comment by yahonda** [comment]
> This was a bug that has been fixed in libvips 8.13 - see:
https://www.libvips.org/2022/05/28/What's-new-in-8.13.html#image-resize-quality-improvements

That totally makes sense because `libvips42/jammy,now 8.12.1-1build1 arm64 [installed]` is installed. Thank for the useful information. Closing this issue.
